sdk/core/Azure.Core.Experimental/src/DynamicData.cs

```diff
@@ -192,6 +192,96 @@ public void Dispose()
             _element.DisposeRoot();
         }
 
+        /// <inheritdoc/>
+        public override bool Equals(object? obj)
+        {
+            if (obj is null)
+            {
+                return _element.ValueKind == JsonValueKind.Null;
+            }
+
+            if (_element.ValueKind == JsonValueKind.Null)
+            {
+                return obj is null || obj is DynamicData d && Equals(d);
+            }
+
+            return obj switch
+            {
+                string s =>
+                    _element.ValueKind == JsonValueKind.String &&
+                    _element.GetString() == s,
+
+                bool b =>
+                    (_element.ValueKind == JsonValueKind.True ||
+                     _element.ValueKind == JsonValueKind.False) &&
+                    _element.GetBoolean() == b,
+
+                double d =>
+                    _element.ValueKind == JsonValueKind.Number &&
+                    _element.TryGetDouble(out double od) && d == od,
+
+                float f =>
+                    _element.ValueKind == JsonValueKind.Number &&
+                    _element.TryGetSingle(out float of) && f == of,
+
+                long l =>
+                    _element.ValueKind == JsonValueKind.Number &&
+                    _element.TryGetInt64(out long ol) && l == ol,
+
+                int i =>
+                    _element.ValueKind == JsonValueKind.Number &&
+                    _element.TryGetInt32(out int oi) && i == oi,
+
+                DynamicData data => Equals(data),
+
+                _ => base.Equals(obj),
+            };
+        }
+
+        internal bool Equals(DynamicData other)
+        {
+            if (other is null)
+            {
+                return _element.ValueKind == JsonValueKind.Null;
+            }
+
+            if (_element.ValueKind != other._element.ValueKind)
+            {
+                return false;
+            }
+
+            return _element.ValueKind switch
+            {
+                JsonValueKind.String => _element.GetString() == other._element.GetString(),
+                JsonValueKind.Number => NumberEqual(other),
+                JsonValueKind.True => true,
+                JsonValueKind.False => true,
+                JsonValueKind.Null => true,
```

Comment:
Shouldn't these compare against `other._element.ValueKind`?

Reply:
I believe that is covered in the line above that returns false if the this/other ValueKinds don't match -- but tell me if you think my reasoning is incorrect here.

Reply:
You're right. My brain just wasn't processing it earlier. I know it was a combination of those two conditions but, well, TGIF.